# Customer Segmentation

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
#parse the dates while reading the csv
customer_data = pd.read_csv('../data/DM_AIAI_CustomerDB.csv', parse_dates=[ 'EnrollmentDateOpening', 'CancellationDate'])

In [4]:
customer_data.head()

Unnamed: 0  Loyalty# First Name    Last Name        Customer Name Country  \
0           0    480934    Cecilia  Householder  Cecilia Householder  Canada   
1           1    549612      Dayle        Menez          Dayle Menez  Canada   
2           2    429460     Necole       Hannon        Necole Hannon  Canada   
3           3    608370      Queen        Hagee          Queen Hagee  Canada   
4           4    530508     Claire      Latting       Claire Latting  Canada   

  Province or State       City   Latitude   Longitude  ...  Gender Education  \
0           Ontario    Toronto  43.653225  -79.383186  ...  female  Bachelor   
1           Alberta   Edmonton  53.544388 -113.490930  ...    male   College   
2  British Columbia  Vancouver  49.282730 -123.120740  ...    male   College   
3           Ontario    Toronto  43.653225  -79.383186  ...    male   College   
4            Quebec       Hull  45.428730  -75.713364  ...    male  Bachelor   

  Location Code   Income  Marital Status LoyaltyStatus EnrollmentDateOpening  \
0         Urban  70146.0         Married          Star            2019-02-15   
1         Rural      0.0        Divorced          Star            2019-03-09   
2         Urban      0.0          Single          Star            2017-07-14   
3      Suburban      0.0          Single          Star            2016-02-17   
4      Suburban  97832.0         Married          Star            2017-10-25   

  CancellationDate Customer Lifetime Value  EnrollmentType  
0              NaN                 3839.14        Standard  
1              NaN                 3839.61        Standard  
2         1/8/2021                 3839.75        Standard  
3              NaN                 3839.75        Standard  
4              NaN                 3842.79  2021 Promotion  

[5 rows x 21 columns]

In [5]:
customer_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16921 entries, 0 to 16920
Data columns (total 21 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   Unnamed: 0               16921 non-null  int64         
 1   Loyalty#                 16921 non-null  int64         
 2   First Name               16921 non-null  object        
 3   Last Name                16921 non-null  object        
 4   Customer Name            16921 non-null  object        
 5   Country                  16921 non-null  object        
 6   Province or State        16921 non-null  object        
 7   City                     16921 non-null  object        
 8   Latitude                 16921 non-null  float64       
 9   Longitude                16921 non-null  float64       
 10  Postal code              16921 non-null  object        
 11  Gender                   16921 non-null  object        
 12  Education                16921 n

- Unnamed 0 is just an index column, can be dropped

In [6]:
customer_data = customer_data.drop(columns=['Unnamed: 0'])
customer_data.head()

Loyalty# First Name    Last Name        Customer Name Country  \
0    480934    Cecilia  Householder  Cecilia Householder  Canada   
1    549612      Dayle        Menez          Dayle Menez  Canada   
2    429460     Necole       Hannon        Necole Hannon  Canada   
3    608370      Queen        Hagee          Queen Hagee  Canada   
4    530508     Claire      Latting       Claire Latting  Canada   

  Province or State       City   Latitude   Longitude Postal code  Gender  \
0           Ontario    Toronto  43.653225  -79.383186     M2Z 4K1  female   
1           Alberta   Edmonton  53.544388 -113.490930     T3G 6Y6    male   
2  British Columbia  Vancouver  49.282730 -123.120740     V6E 3D9    male   
3           Ontario    Toronto  43.653225  -79.383186     P1W 1K4    male   
4            Quebec       Hull  45.428730  -75.713364     J8Y 3Z5    male   

  Education Location Code   Income Marital Status LoyaltyStatus  \
0  Bachelor         Urban  70146.0        Married          Star   
1   College         Rural      0.0       Divorced          Star   
2   College         Urban      0.0         Single          Star   
3   College      Suburban      0.0         Single          Star   
4  Bachelor      Suburban  97832.0        Married          Star   

  EnrollmentDateOpening CancellationDate  Customer Lifetime Value  \
0            2019-02-15              NaN                  3839.14   
1            2019-03-09              NaN                  3839.61   
2            2017-07-14         1/8/2021                  3839.75   
3            2016-02-17              NaN                  3839.75   
4            2017-10-25              NaN                  3842.79   

   EnrollmentType  
0        Standard  
1        Standard  
2        Standard  
3        Standard  
4  2021 Promotion

In [7]:
customer_data.describe()

Loyalty#      Latitude     Longitude        Income  \
count   16921.000000  16921.000000  16921.000000  16901.000000   
mean   550197.393771     47.174500    -91.814768  37758.038400   
min    100011.000000     42.984924   -135.056840      0.000000   
25%    326823.000000     44.231171   -120.237660      0.000000   
50%    550896.000000     46.087818    -79.383186  34161.000000   
75%    772438.000000     49.282730    -74.596184  62396.000000   
max    999999.000000     60.721188    -52.712578  99981.000000   
std    259251.503597      3.307971     22.242429  30368.992499   

               EnrollmentDateOpening  Customer Lifetime Value  
count                          16921             16901.000000  
mean   2018-10-05 09:50:05.567046912              7990.460188  
min              2015-01-27 00:00:00              1898.010000  
25%              2017-01-17 00:00:00              3979.720000  
50%              2018-10-31 00:00:00              5780.180000  
75%              2020-07-09 00:00:00              8945.690000  
max              2021-12-30 00:00:00             83325.380000  
std                              NaN              6863.173093

- Loyalty is just a unique identifier and longitude and latitude may not be useful for clustering.
Income and Customer Lifetime Value are the important features for segmentation.

In [8]:
customer_data.describe(include=['object'])

First Name Last Name Customer Name Country Province or State     City  \
count       16921     16921         16921   16921             16921    16921   
unique       4941     15404         16921       1                11       29   
top          Deon   Salberg   Emma Martin  Canada           Ontario  Toronto   
freq           13         4             1   16921              5468     3390   

       Postal code  Gender Education Location Code Marital Status  \
count        16921   16921     16921         16921          16921   
unique          75       2         5             3              3   
top        V6E 3D9  female  Bachelor      Suburban        Married   
freq           917    8497     10586          5716           9842   

       LoyaltyStatus CancellationDate EnrollmentType  
count          16921             2310          16921  
unique             3             1260              2  
top             Star         7/7/2020       Standard  
freq            7761                8          15773

- The dataset is of only one country - Canada.
- The Country column can be dropped as it has no variation and won't contribute to clustering.

In [9]:
customer_data = customer_data.drop(columns=['Country'])

In [10]:
customer_data['FullName'] = customer_data['First Name'] + ' ' + customer_data['Last Name']
customer_data[['Customer Name', 'FullName']].head()

Customer Name             FullName
0  Cecilia Householder  Cecilia Householder
1          Dayle Menez          Dayle Menez
2        Necole Hannon        Necole Hannon
3          Queen Hagee          Queen Hagee
4       Claire Latting       Claire Latting

In [11]:
(customer_data['Customer Name'] != customer_data['FullName']).sum()

np.int64(0)

- We can see that the first name and last name are present along with the customer name so we can drop those columns as they are redundant.

In [12]:
customer_data = customer_data.drop(columns=['Customer Name', 'First Name', 'Last Name'])

In [13]:
customer_data.Gender.value_counts()

Gender
female    8497
male      8424
Name: count, dtype: int64

- Male and female distribution is almost equal.

In [14]:
customer_data.Education.value_counts()

Education
Bachelor                10586
College                  4283
High School or Below      792
Doctor                    742
Master                    518
Name: count, dtype: int64

- Most of them have the Bachelors Degree.

In [15]:
customer_data['Location Code'].value_counts()

Location Code
Suburban    5716
Rural       5668
Urban       5537
Name: count, dtype: int64

In [16]:
customer_data['LoyaltyStatus'].value_counts()

LoyaltyStatus
Star      7761
Nova      5722
Aurora    3438
Name: count, dtype: int64

In [17]:
customer_data.isna().sum()

Loyalty#                       0
Province or State              0
City                           0
Latitude                       0
Longitude                      0
Postal code                    0
Gender                         0
Education                      0
Location Code                  0
Income                        20
Marital Status                 0
LoyaltyStatus                  0
EnrollmentDateOpening          0
CancellationDate           14611
Customer Lifetime Value       20
EnrollmentType                 0
FullName                       0
dtype: int64

- Income and CLV has 20 missing values and CacellationDate has 14611 missing values.

In [19]:
customer_data.isna().sum() / len(customer_data) * 100

Loyalty#                    0.000000
Province or State           0.000000
City                        0.000000
Latitude                    0.000000
Longitude                   0.000000
Postal code                 0.000000
Gender                      0.000000
Education                   0.000000
Location Code               0.000000
Income                      0.118196
Marital Status              0.000000
LoyaltyStatus               0.000000
EnrollmentDateOpening       0.000000
CancellationDate           86.348325
Customer Lifetime Value     0.118196
EnrollmentType              0.000000
FullName                    0.000000
dtype: float64

In [ ]:
- About 86.25% of the CancellationDate values are missing which indicates that most customers are still active.

In [20]:
customer_data['Province or State'].value_counts()

Province or State
Ontario                 5468
British Columbia        4416
Quebec                  3311
Alberta                 1008
Manitoba                 676
New Brunswick            652
Nova Scotia              541
Saskatchewan             412
Newfoundland             258
Yukon                    112
Prince Edward Island      67
Name: count, dtype: int64

In [21]:
customer_data['Marital Status'].value_counts()

Marital Status
Married     9842
Single      4539
Divorced    2540
Name: count, dtype: int64

In [24]:
customer_data['year'] = customer_data['EnrollmentDateOpening'].dt.year
customer_data['month'] = customer_data['EnrollmentDateOpening'].dt.month
customer_data['day'] = customer_data['EnrollmentDateOpening'].dt.day

In [23]:
customer_data['year'].value_counts()

year
2021    3033
2020    2519
2019    2477
2016    2420
2017    2407
2018    2351
2015    1714
Name: count, dtype: int64

- They have been enrolled from 2015 to 2021 with maximum enrollments in 2021.

- As we have extracted year, month and day from EnrollmentDateOpening, we can drop the original column now.

In [26]:
customer_data = customer_data.drop(columns=['EnrollmentDateOpening'])

In [28]:
customer_data[customer_data['Income'].isna()]

Loyalty# Province or State          City   Latitude   Longitude  \
16901    999987     New Brunswick   Fredericton  46.029263  -66.565150   
16902    999988            Quebec      Montreal  45.573672  -73.523012   
16903    999989           Ontario       Trenton  44.075379  -77.550375   
16904    999990     New Brunswick       Moncton  46.106617  -64.714267   
16905    999991     New Brunswick   Fredericton  45.950000  -66.652437   
16906    999992           Ontario       Toronto  43.706878  -79.437412   
16907    999993  British Columbia  Dawson Creek  55.701475 -120.181716   
16908    999994           Ontario        Ottawa  45.365906  -75.723181   
16909    999995           Ontario        Ottawa  45.471557  -75.704868   
16910    999996           Ontario       Toronto  43.690489  -79.436758   
16911    999997     New Brunswick       Moncton  46.067243  -64.729480   
16912    999998            Quebec   Quebec City  46.749897  -71.238050   
16913    999999           Alberta      Edmonton  53.549353 -113.509109   
16914    100017  British Columbia  Dawson Creek  55.751178 -120.264920   
16915    100011           Ontario       Toronto  43.593187  -79.444335   
16916    100012            Quebec   Quebec City  46.759733  -71.141009   
16917    100013           Alberta      Edmonton  53.524829 -113.546357   
16918    100014     New Brunswick       Moncton  46.051866  -64.825428   
16919    100015            Quebec   Quebec City  46.862970  -71.133444   
16920    100016  British Columbia  Dawson Creek  55.720562 -120.160090   

      Postal code  Gender Education Location Code  Income Marital Status  \
16901     R4H 2Y2  female  Bachelor         Urban     NaN         Single   
16902     N6B 1N3    male   College         Rural     NaN         Single   
16903     P8F 5C8    male   College         Rural     NaN        Married   
16904     B6P 6D0  female   College         Rural     NaN        Married   
16905     X3W 5N2  female   College      Suburban     NaN        Married   
16906     P6D 6N2    male   College      Suburban     NaN         Single   
16907     W6H 0Z7  female   College      Suburban     NaN        Married   
16908     B2F 3E1  female   College         Rural     NaN        Married   
16909     B3A 2R0  female   College      Suburban     NaN        Married   
16910     B4W 4M6  female  Bachelor      Suburban     NaN         Single   
16911     S5R 0X5    male   College         Urban     NaN        Married   
16912     K7T 9I9    male  Bachelor         Urban     NaN         Single   
16913     S7D 0V1    male  Bachelor         Rural     NaN         Single   
16914     E0K 5I2    male    Master         Urban     NaN        Married   
16915     W9D 4Q9  female  Bachelor      Suburban     NaN        Married   
16916     Y0C 7D6    male  Bachelor      Suburban     NaN         Single   
16917     L3S 9Y3  female  Bachelor         Rural     NaN        Married   
16918     G2S 2B6    male  Bachelor         Rural     NaN        Married   
16919     B1Z 8T3  female   College         Urban     NaN        Married   
16920     M4A 1E4  female    Master      Suburban     NaN         Single   

      LoyaltyStatus CancellationDate  Customer Lifetime Value EnrollmentType  \
16901          Star         3/7/2017                      NaN       Standard   
16902          Star        8/22/2017                      NaN       Standard   
16903          Star        9/12/2015                      NaN       Standard   
16904          Star        6/10/2019                      NaN       Standard   
16905          Star        7/20/2019                      NaN       Standard   
16906          Star        3/27/2021                      NaN       Standard   
16907          Star        1/27/2015                      NaN       Standard   
16908          Star         5/5/2016                      NaN       Standard   
16909          Star         3/2/2020                      NaN       Standard   
16910          Star        9/14/2018                   

- We can see that the 20 missing Income values correspond to customers with missing Customer Lifetime Value.

In [29]:
customer_data.duplicated().sum()

np.int64(0)

- No duplicate rows found in the dataset.

In [ ]:
customer_data.to_csv('../data/customer.csv', index=False)